posts/Stef/2019-11-18-hibernate-orm-with-panache-in-quarkus.adoc

```diff
@@ -0,0 +1,465 @@
+        personDao.delete(p);
+    }
+    
+    @POST
+    @Path("people")
+    public Response newPerson(@Context UriInfo uriInfo, Person newPerson){
+        Person p = new Person();
+        p.setBirth(newPerson.getBirth());
+        p.setFirstName(newPerson.getFirstName());
+        p.setLastName(newPerson.getLastName());
+        personDao.persist(p);
+        
+        URI uri = uriInfo.getAbsolutePathBuilder()
+                .path(PersonEndpoint.class)
+                .path(PersonEndpoint.class, "findById")
+                .build(p.getId());
+        return Response.created(uri).build();
+    }
+}
+----
+
+
+== A few observations on the traditional JPA way
+
+We've all seen hundreds of entities and DAOs written this way. They don't have anything surprising.
+
+What they do have plenty of, on the other hand is boilerplate:
+
+- The generated ID field. Very often, all your entities will use the same auto-generate ID type.
+- All those property accessors that do nothing in the entity. They are required for encapsulation, and because Java
+does not support first-class properties in the language. Most people either generate them from their IDE, or using
+https://projectlombok.org/features/GetterSetter[Lombok].
+- All those `persist`/`delete`/`findById`/`findAll` methods on every DAO. All DAOs have them.
+- Those DAO queries all start with `FROM Person` and have to repeat the `Person.class` all over the place.
+
+== A first look at what Hibernate ORM with Panache can do for us
+
+Let's jump forward to https://quarkus.io/[Quarkus] and in particular, what 
+https://quarkus.io/guides/hibernate-orm-panache[Hibernate ORM with Panache] can do for us. It turns out to be quite a lot.
+
+Quarkus allows us to do a lot of bytecode modification at build-time, which (among many benefits) lets us side-step
+Java's lack of support for first-class properties by:
+
+- Writing public fields instead of private+getter+setter
+- Hibernate ORM with Panache will actually generate any getter+setter that you did not write, and
+- It will replace all field accesses with accesses to the getters and setters.
+
+This system allows us to write code as if we were using public fields, but behind the scenes, we still get encapsulation
+and forward-compatibility if we ever add getters or setters that do more than just access the field.
+
+On top of that, Hibernate ORM with Panache comes with support for DAOs that already have a lot of the methods that
+you commonly write.
+
+We can thus rewrite our previous entity class by extending `PanacheEntity` which comes with a predefined auto-generated
+ID field:
+
+[source%nowrap,java]
+----
+@Entity
+public class Person extends PanacheEntity {
+
+    public String firstName;
+    public String lastName;
+    public Date birth;
+}
+----
+
+And we can rewrite our DAO by extending `PanacheRepository` to get all the common methods:
+
+[source%nowrap,java]
+----
+@Singleton
+public class PersonDao implements PanacheRepository<Person>{
```

Comment:
could you have a white space before the `{` ? it looks so bad :)

Same on the following lines...

Reply:
OK